docs/architecture/data-sources.md

```diff
@@ -0,0 +1,113 @@
+# Data Sources
+
+Data sources allow developers easily integrate different data providers into the carbon aware SDK ([WattTime](https://www.wattime.org), [ElectricityMap](https://static.electricitymap.org), etc) to be made available to all higher level user-interfaces (WebAPI, CLI, etc), while avoiding the details of how to interact with any specific provider.
+
+## Data Sources' Responsibility
+Data sources act as the data ingestion layer for the SDK, handling the retrieval of data from a given data provider. They contain specific knowledge about the data provider they access, such as flags used in requests, fields that come back in responses, special use cases etc. They also handle any external calls that must be made to access the data provider. While helper clients can be built to handle the these calls, only the data source should have access to, and knowledge of, that client.
+- For example, the WattTimeDataSource has a reference to a private WattTimeClient within it's implementation. The WattTimeClient handles the HTTP GET/POST calls to WattTime and the data source invokes the client once it has processed the request, and then processes the response before returning a final result.
+
+### Aggregator <-> Data Source Contract
+In order for the SDK to support different data sources, there is a defined contract between the Aggregator layer and the DataSource layer. This contract defines what a request from the Aggregator to the DataSource must contain, and what type of response the Aggregator expects from the DataSource. This means that each data source is responsible for: 
+- Pre-processing the Aggregator request must to create the expected request for the data provider.
+- Post-processing the data provider result to create the expected response type for the Aggregator
+
+### Post-Processing Caveat
```

Comment:
```suggestion
#### Post-Processing Caveat
```

Consider nesting this section within the above section since they seem related in that way more than as siblings